### PR TITLE
Fix intrinsic applied for array slices and stored in array slice targets

### PR DESCRIPTION
**Pull request by czgdp1807**
https://github.com/certik/fastGPT/pull/55#issuecomment-1684217628

### CODE REVIEW

**Comment by czgdp1807** [comment]
Works totally fine on my Apple M2 Pro macbook,

```zsh
(fastgptlf) 18:32:15:~/lfortran_project/fastGPT/lf % ln -s ../../model_fastgpt_124M_v1.dat model.dat
(fastgptlf) 18:32:19:~/lfortran_project/fastGPT/lf % FC=lfortran cmake -DFASTGPT_BLAS=OpenBLAS ..
-- The C compiler identification is AppleClang 14.0.3.14030022
-- The CXX compiler identification is AppleClang 14.0.3.14030022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The Fortran compiler identification is unknown
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - failed
-- Check for working Fortran compiler: /Users/czgdp1807/lfortran_project/lfortran/src/bin/lfortran
-- Check for working Fortran compiler: /Users/czgdp1807/lfortran_project/lfortran/src/bin/lfortran - works
-- Checking whether /Users/czgdp1807/lfortran_project/lfortran/src/bin/lfortran supports Fortran 90
-- Checking whether /Users/czgdp1807/lfortran_project/lfortran/src/bin/lfortran supports Fortran 90 - no
-- Found OPENBLAS: /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/System/Library/Frameworks/vecLib.framework/Headers  
-- Found OMP: /Users/czgdp1807/mambaforge/envs/fastgptlf/include  


Configuration results
---------------------
Fortran compiler: /Users/czgdp1807/lfortran_project/lfortran/src/bin/lfortran
Build type: Release
Fortran compiler flags: 
Installation prefix: /usr/local
FASTGPT_BLAS: OpenBLAS
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/czgdp1807/lfortran_project/fastGPT/lf
(fastgptlf) 18:32:23:~/lfortran_project/fastGPT/lf % make
[  5%] Building Fortran object CMakeFiles/fastgpt.dir/linalg_c.f90.o
[ 11%] Building Fortran object CMakeFiles/fastgpt.dir/tokenizer.f90.o
[ 17%] Building Fortran object CMakeFiles/fastgpt.dir/gpt2.f90.o
[ 23%] Building Fortran object CMakeFiles/fastgpt.dir/omp.f90.o
[ 29%] Building Fortran object CMakeFiles/fastgpt.dir/driver.f90.o
[ 35%] Building C object CMakeFiles/fastgpt.dir/linalg_openblas.c.o
[ 41%] Linking Fortran static library libfastgpt.a
[ 41%] Built target fastgpt
[ 47%] Building Fortran object CMakeFiles/gpt2.dir/main.f90.o
[ 52%] Linking Fortran executable gpt2
[ 52%] Built target gpt2
[ 58%] Building Fortran object CMakeFiles/chat.dir/chat.f90.o
warning: format string in read() is not implemented yet and it is currently treated as '*'
 --> /Users/czgdp1807/lfortran_project/fastGPT/chat.f90:1:1
  |
1 | program chatgpt2
  | ^ treated as '*'


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
[ 64%] Linking Fortran executable chat
[ 64%] Built target chat
[ 70%] Building Fortran object CMakeFiles/test_basic_input.dir/tests/test_basic_input.f90.o
[ 76%] Linking Fortran executable test_basic_input
[ 76%] Built target test_basic_input
[ 82%] Building Fortran object CMakeFiles/test_more_inputs.dir/tests/test_more_inputs.f90.o
[ 88%] Linking Fortran executable test_more_inputs
[ 88%] Built target test_more_inputs
[ 94%] Building Fortran object CMakeFiles/test_chat.dir/tests/test_chat.f90.o
warning: format string in read() is not implemented yet and it is currently treated as '*'
 --> /Users/czgdp1807/lfortran_project/fastGPT/tests/test_chat.f90:1:1
  |
1 | program test_chat
  | ^ treated as '*'


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
[100%] Linking Fortran executable test_chat
[100%] Built target test_chat
(fastgptlf) 18:32:30:~/lfortran_project/fastGPT/lf % ./gpt2
Loading the model...
    done. Time:.112E+00s, Model file version: 1

Model parameters:
n_vocab = 50257
n_ctx   =  1024
n_embd  =   768
n_layer =    12
n_head  =    12

Input text
Alan Turing theorized that computers would one day become very powerful, but even he could not imagine

Encoding: tokenizing input text into tokens (currently slow)...
    done. Time:.258E+00s

Input parameters:
n_seq                =  19
n_tokens_to_generate =  20

Input tokens:
36235 39141 18765 1143 326 9061 561 530 1110 1716 845 3665 11 475 772 339 714 407 5967 

Decoded input as text:
Alan Turing theorized that computers would one day become very powerful, but even he could not imagine

Running model....
 how
 they
 would
 be
 able
 to
 do
 so
.




"
I
 think
 that
 the
 most
 important
 thing
 is

    done. Time:.184E+01s (0.74E+01x)

Output tokens:
703 484 561 307 1498 284 466 523 13 198 198 1 40 892 326 262 749 1593 1517 318 

Decoded output as text:
 how they would be able to do so.

"I think that the most important thing is
(fastgptlf) 18:32:34:~/lfortran_project/fastGPT/lf
```

I will see what's different happening on Linux.

**Comment by czgdp1807** [comment]
Now I can't reproduce this in my Linux system as well. It only fails on CI. Works on both my mac and linux.

```bash
(fastgpt) czgdp1807ssd@czgdp1807ssd:~/lfortran_project/fastGPT/lf$ FC=lfortran CMAKE_PREFIX_PATH=$CONDA_PREFIX cmake -DFASTGPT_BLAS=OpenBLAS ..
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The Fortran compiler identification is unknown
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - failed
-- Check for working Fortran compiler: /home/czgdp1807ssd/lfortran_project/lfortran/src/bin/lfortran
-- Check for working Fortran compiler: /home/czgdp1807ssd/lfortran_project/lfortran/src/bin/lfortran - works
-- Checking whether /home/czgdp1807ssd/lfortran_project/lfortran/src/bin/lfortran supports Fortran 90
-- Checking whether /home/czgdp1807ssd/lfortran_project/lfortran/src/bin/lfortran supports Fortran 90 - no
-- Found OPENBLAS: /home/czgdp1807ssd/miniforge3/envs/fastgpt/include  
-- Found OMP: /usr/include  


Configuration results
---------------------
Fortran compiler: /home/czgdp1807ssd/lfortran_project/lfortran/src/bin/lfortran
Build type: Release
Fortran compiler flags: 
Installation prefix: /usr/local
FASTGPT_BLAS: OpenBLAS
-- Configuring done
-- Generating done
-- Build files have been written to: /home/czgdp1807ssd/lfortran_project/fastGPT/lf
(fastgpt) czgdp1807ssd@czgdp1807ssd:~/lfortran_project/fastGPT/lf$ make gpt2
[ 11%] Building Fortran object CMakeFiles/fastgpt.dir/linalg_c.f90.o
[ 22%] Building Fortran object CMakeFiles/fastgpt.dir/tokenizer.f90.o
[ 33%] Building Fortran object CMakeFiles/fastgpt.dir/gpt2.f90.o
[ 44%] Building Fortran object CMakeFiles/fastgpt.dir/omp.f90.o
[ 55%] Building Fortran object CMakeFiles/fastgpt.dir/driver.f90.o
[ 66%] Building C object CMakeFiles/fastgpt.dir/linalg_openblas.c.o
[ 77%] Linking Fortran static library libfastgpt.a
[ 77%] Built target fastgpt
[ 88%] Building Fortran object CMakeFiles/gpt2.dir/main.f90.o
[100%] Linking Fortran executable gpt2
[100%] Built target gpt2
(fastgpt) czgdp1807ssd@czgdp1807ssd:~/lfortran_project/fastGPT/lf$ ./gpt2
./gpt2: Relink `/home/czgdp1807ssd/miniforge3/envs/fastgpt/lib/./libgfortran.so.5' with `/lib/x86_64-linux-gnu/librt.so.1' for IFUNC symbol `clock_gettime'
Loading the model...
    done. Time:.117E+01s, Model file version: 1

Model parameters:
n_vocab = 50257
n_ctx   =  1024
n_embd  =   768
n_layer =    12
n_head  =    12

Input text
Alan Turing theorized that computers would one day become very powerful, but even he could not imagine

Encoding: tokenizing input text into tokens (currently slow)...
    done. Time:.750E+00s

Input parameters:
n_seq                =  19
n_tokens_to_generate =  20

Input tokens:
36235 39141 18765 1143 326 9061 561 530 1110 1716 845 3665 11 475 772 339 714 407 5967 

Decoded input as text:
Alan Turing theorized that computers would one day become very powerful, but even he could not imagine

Running model....
 how
 they
 would
 be
 able
 to
 do
 so
.




"
I
 think
 that
 the
 most
 important
 thing
 is

    done. Time:.196E+01s (0.80E+01x)

Output tokens:
703 484 561 307 1498 284 466 523 13 198 198 1 40 892 326 262 749 1593 1517 318 

Decoded output as text:
 how they would be able to do so.

"I think that the most important thing is
```

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -1410,6 +1410,7 @@ LFORTRAN_API void _lfortran_free(char* ptr) {
 
 LFORTRAN_API void _lfortran_alloc(char** ptr, int32_t size) {
     *ptr = (char *) malloc(size);
+    memset(*ptr, '\0', size);
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:1413]: We should not be initializing the memory to 0.

**Comment by certik** [comment]
The CI in master seems to work and not exhibit this bug.

But the PR https://github.com/lfortran/lfortran/pull/2216 triggers this bug.

If you do not modify the fastgpt test at all, just implement the feature, it still triggers it?

If not, get the PR ready to merge, let's merge. Then we'll worry about fastgpt. If yes, then we need to make this PR smaller, until we get something that can be merged.

**Comment by czgdp1807** [comment]
The bug is now fixed. There are two reasons,

1. Garbage values after allocating a string using `malloc`. In macOS, this issue doesn't happen, however in linux this does happen because of the garbage values in character block allocated in heap memory. So, I solved this issue by filling the whole string with `'\0'`. It only happens for strings. See the changes in `asr_to_llvm.cpp`. It is necessary to do because if we don't then `len(string)` shows flaky behaviour and sometimes gives correct output and sometimes doesn't. I have faced this flaky behaviour on my linux system some times. 
2.  Also, I am allocating `size(x) + 1` in `c2s`. That ` + 1` is for storing the `\0` character at the end. If you don't then again `len(string)` gives flaky behaviour on CI and sometimes on my linux system.

It isn't at all related to what changes I made. macOS is robust so my PR always worked on my Apple M2 Pro Macbook even before I did the above 2 changes. Its only a problem with linux and that because of garbage values. See some excerpts from logs,

```bash
2023-08-25T10:18:23.0878463Z Running model....
2023-08-25T10:18:23.0878582Z y.before:  0  5 5
2023-08-25T10:18:23.0878715Z 1 1 � -60
2023-08-25T10:18:23.0878834Z 2 2 Ġ -96
2023-08-25T10:18:23.0878958Z 3 3 Ġh 104
2023-08-25T10:18:23.0879081Z 4 4 Ġho 111
2023-08-25T10:18:23.0879209Z 5 5 Ġhow 119
2023-08-25T10:18:23.0879348Z y.after:  5 Ġhow
2023-08-25T10:18:23.0879433Z  how
2023-08-25T10:18:23.0879515Z y.before:  0  6 6
2023-08-25T10:18:23.0879636Z 1 1 � -60
2023-08-25T10:18:23.0879751Z 2 2 Ġ -96
2023-08-25T10:18:23.0879871Z 3 3 Ġt 116
2023-08-25T10:18:23.0879992Z 4 4 Ġth 104
2023-08-25T10:18:23.0880116Z 5 5 Ġthe 101
2023-08-25T10:18:23.0880309Z 6 25 Ġthey�?��>��Ⱦ}wF?24`?! 121
2023-08-25T10:18:23.0880531Z y.after:  25 Ġthey�?��>��Ⱦ}wF?24`?!
2023-08-25T10:18:23.0880697Z Codepoint out of range for byte decoder: 767 323
```

You can clearly see how `2023-08-25T10:18:23.0880309Z 6 25 Ġthey�?��>��Ⱦ}wF?24`?! 121` happens. That's because of the point 2 I mentioned above. There is no space for `null` character at the end so `len(y)` gives 25. Now that doesn't happen.

**Bottomline** - The above is the best I could do to resolve the flaky behaviour and it is only related to the behaviour of strings and not array computations in `gpt2.f90`.

**Comment by certik** [comment]
I've seen this flaky behavior too. TODO:
* [ ] send a separate PR fixing the flaky behavior
* [ ] let's finish this PR
* [ ] unrelated: fix the scipy failure, in a separeate PR, see this comment how to fix it: https://github.com/lfortran/lfortran/pull/2243#issuecomment-1693566108

**Comment by czgdp1807** [comment]
I already fixed the flaky behaviour in this PR. That's the best I could do. Nothing more to be done from end in this PR. Please review and let me know if it looks fine to you.

**Review by certik — APPROVED** [review]
I think this is good!

**Review by certik — APPROVED** [review]

**Pull request merged**